Cylindrical prism with optional filleted edges

yardasol opened the pull request:
This PR adds a new function, `cylindrical_prism`, that creates a finite cylindrical prism region with the option to fillet the cylinder's edges

gonuke commented on openmc/model/funcs.py:
```diff
@@ -110,6 +111,140 @@ def borated_water(boron_ppm, temperature=293., pressure=0.1013, temp_unit='K',
+    Returns
+    -------
+    openmc.Region
+        The inside of a rectangular prism
```
This looks like a cut/paste typo

gonuke commented on openmc/model/funcs.py:
```diff
@@ -110,6 +111,140 @@ def borated_water(boron_ppm, temperature=293., pressure=0.1013, temp_unit='K',
+                      boundary_type='transmission',
+                      upper_edge_radius=0.,
+                      lower_edge_radius=0.):
+    """Get a finite cylindrical prism.
```
Perhaps more information here about the rounded edges.  From reviewing the code, it seems that this option adds what I would call a fillet at each end.

pshriwise replied:
+1 for fillet terminology

gonuke commented on openmc/model/funcs.py:
```diff
@@ -110,6 +111,140 @@ def borated_water(boron_ppm, temperature=293., pressure=0.1013, temp_unit='K',
+        axcoord1 = 0
+        axcoord2 = 1
+
+    # Get torus class corresponding to given axis
+    tor = globals()['{}Torus'.format(axis.upper())]
```
You only need this if you are implementing the rounded edge/fillet, so perhaps make it conditional and/or move it closer to that code section.

gonuke commented on openmc/model/funcs.py:
```diff
@@ -110,6 +111,140 @@ def borated_water(boron_ppm, temperature=293., pressure=0.1013, temp_unit='K',
+        args[x1 + '0'] = origin[axcoord1]
+        args[x2 + '0'] = origin[axcoord2]
+        if upper_edge_radius > 0.:
+            args['a'] = r - upper_edge_radius
+            args['b'] = upper_edge_radius
+            args['c'] = upper_edge_radius
+            args[axis + '0'] = origin[axcoord] + height/2 - upper_edge_radius
+            tor_upper_max = tor(name='{} max'.format(axis), **args)
+
+            axis_upper_min = plane(axis, 'upper min', height/2 + origin[axcoord] - upper_edge_radius)
+            cyl_upper_min = cylinder(axis, 'upper min', x1, origin[axcoord1], x2, origin[axcoord2], r - upper_edge_radius)
+
+            corners_upper = (+cyl_upper_min & +tor_upper_max & + axis_upper_min)
+
+        if lower_edge_radius > 0.:
+            args['a'] = r - lower_edge_radius
+            args['b'] = lower_edge_radius
+            args['c'] = lower_edge_radius
+            args[axis + '0'] = origin[axcoord] - height/2 + lower_edge_radius
+            tor_lower_min = tor(name='{} min'.format(axis), **args)
+
+            axis_lower_max = plane(axis, 'lower max', origin[axcoord] - height/2 + lower_edge_radius)
+            cyl_lower_min = cylinder(axis, 'lower min', x1, origin[axcoord1], x2, origin[axcoord2], r - lower_edge_radius)
+
+            corners_lower = (+cyl_lower_min & +tor_lower_min & - axis_lower_max)
+
```
It seems as though there is an opportunity to reuse at least some of this code in a modular way (e.g. a function or loop) rather than cut/paste for the upper/lower.  The benefit of this kind of reuse is for consistency/maintainability and not for performance.

gonuke reviewed (changes requested):
This looks like a useful addition @yardasol.  I have a few suggestions related mostly to documentation and code reuse/maintainability.

paulromano commented on openmc/model/funcs.py:
```diff
@@ -110,6 +111,140 @@ def borated_water(boron_ppm, temperature=293., pressure=0.1013, temp_unit='K',
+                   boundary_type=boundary_type,
+                   **{axis + '0': value})
+    def cylinder(axis, name, ax1, val1, ax2, val2, r):
+        cls = globals()['{}Cylinder'.format(axis.upper())]
+        return cls(name='{} {}'.format(name, axis),
```
```suggestion
        cls = getattr(openmc, f'{axis.upper()}Cylinder')
        return cls(name=f'{name} {axis}',
```

paulromano commented on openmc/model/funcs.py:
```diff
@@ -110,6 +111,140 @@ def borated_water(boron_ppm, temperature=293., pressure=0.1013, temp_unit='K',
+
+        corners_upper = None
+        corners_lower = None
+        args = {'boundary_type': boundary_type}
+        args[x1 + '0'] = origin[axcoord1]
+        args[x2 + '0'] = origin[axcoord2]
```
```suggestion
        args = {
            'boundary_type': boundary_type,
            x1 + '0': origin[axcoord1],
            x2 + '0': origin[axcoord2]
        }
```

paulromano commented on openmc/model/funcs.py:
```diff
@@ -110,6 +111,140 @@ def borated_water(boron_ppm, temperature=293., pressure=0.1013, temp_unit='K',
+            args['b'] = upper_edge_radius
+            args['c'] = upper_edge_radius
+            args[axis + '0'] = origin[axcoord] + height/2 - upper_edge_radius
+            tor_upper_max = tor(name='{} max'.format(axis), **args)
```
```suggestion
            tor_upper_max = tor(name=f'{axis} max', **args)
```

paulromano commented on openmc/model/funcs.py:
```diff
@@ -110,6 +111,140 @@ def borated_water(boron_ppm, temperature=293., pressure=0.1013, temp_unit='K',
+            axis_upper_min = plane(axis, 'upper min', height/2 + origin[axcoord] - upper_edge_radius)
+            cyl_upper_min = cylinder(axis, 'upper min', x1, origin[axcoord1], x2, origin[axcoord2], r - upper_edge_radius)
+
+            corners_upper = (+cyl_upper_min & +tor_upper_max & + axis_upper_min)
```
```suggestion
            corners_upper = +cyl_upper_min & +tor_upper_max & +axis_upper_min
```

paulromano commented on openmc/model/funcs.py:
```diff
@@ -110,6 +111,140 @@ def borated_water(boron_ppm, temperature=293., pressure=0.1013, temp_unit='K',
+            axis_lower_max = plane(axis, 'lower max', origin[axcoord] - height/2 + lower_edge_radius)
+            cyl_lower_min = cylinder(axis, 'lower min', x1, origin[axcoord1], x2, origin[axcoord2], r - lower_edge_radius)
+
+            corners_lower = (+cyl_lower_min & +tor_lower_min & - axis_lower_max)
```
```suggestion
            corners_lower = (+cyl_lower_min & +tor_lower_min & -axis_lower_max)
```

paulromano reviewed (changes requested):
Thanks for this contribution @yardasol! In addition to the line comments, can you also add this function to the documentation in `docs/source/pythonapi/model.rst`?

pshriwise reviewed (changes requested):
Just a few smalll comments from me for now, @yardasol. Sorry it's taken me a bit to get to.

One thought that came to mind is the case where the fillet radius is larger than the cylinder radius. This would result in a degenerate torus (a minor radius larger than the major radius). I _think_ we can do transport on those (@paulromano would know for sure), but I'm curious about your thoughts on what the behavior of this function should be in that case. My initial feeling is that if the corner radius is greater than the cylinder radius, we should limit the fillet radius to the radius of the cylinder and display a warning.

yardasol commented:
Thanks for the reviews everyone! I believe I have addressed everyone's comments.

@pshriwise I added the cell check via `openmc.lib` like you asked, but I don't think it's behaving correctly. The rest of the `test_geometry.py` test seems to get skipped after `test_cylindrical_prism()`. I'll try to take a look at it with a profiler but I'd also like to know your thoughts on my implementation and also on this behavior.

pshriwise commented:
> I believe that I enforce the fillet radii must be less than the cylinder radius using the `check_less_than()` function near the beginning of the function

Ah, yeah so you do. Missed it. Sorry about that.

pshriwise commented on tests/unit_tests/test_geometry.py:
```diff
@@ -203,6 +204,83 @@ def test_get_by_name():
+                       (-2.4, 1, 0.51)]]
+
+    openmc.lib.init()
+    for prism, inside_points, outside_points in zip(prisms, inside_points, outside_points):
```
There are a few things to address here:

  - When `openmc.lib.init` is called, it will read the XML files and initialize OpenMC in memory, so any subsequent XML files that are generated will be ignored. The `init` and `finalize` calls should be inside of the loop so OpenMC is started and shutdown for each prism being tested here. I'd even look into the `openmc.lib.run_in_memory()` context manager too to simplify this further.
  - The name `cell` is going to be undefined in the loop here. Might be good to return the cell created from the `make_geometry` function.
  - Going to need a boundary condition on at least one surface in the geometry.
  - Need to wrap the string passed to `match` in `pytest.raises` in an `re.escape` call to ask the regex matching not to treat the parentheses characters as a pattern. I've been hit by that one before.

yardasol replied:
Okay, got it working!

gonuke commented on openmc/model/funcs.py:
```diff
@@ -111,6 +112,151 @@ def borated_water(boron_ppm, temperature=293., pressure=0.1013, temp_unit='K',
+            x2 + '0': origin[axcoord2]
+        }
+
+        def create_fillet(args, r, fillet_radius, pos):
+            args['a'] = r - fillet_radius
+            args['b'] = fillet_radius
+            args['c'] = fillet_radius
+            if pos:
+                coord = origin[axcoord] + height / 2 - fillet_radius
+                tor_name = f'{axis} max'
+                cyl_name = 'upper_min'
+                axial_bound = +plane(axis, 'upper min', coord)
+            else:
+                coord = origin[axcoord] - height / 2 + lower_fillet_radius
+                tor_name = f'{axis} min'
+                cyl_name = 'lower_min'
+                axial_bound = -plane(axis, 'lower max', coord)
```
I was thinking something like this, but not sure if it's
1. 100% valid (can surface sense be changes as a multiply operation?)
2. too clever? Is it harder to understand for future developers?
3. messing up surface naming conventions?


```suggestion
        def create_fillet(args, r, fillet_radius, pos='upper'):
            args['a'] = r - fillet_radius
            args['b'] = fillet_radius
            args['c'] = fillet_radius
            fillet_ext = height / 2 - fillet_radius
            tor_name = f'{axis} {pos}'
            cyl_name = f'{pos}_min'
            dir = 1
            if pos == 'lower':
                dir = -1
            coord = origin[axcoord] + dir * fillet_ext
            axial_bound = dir * plane(axis, f'{pos} ext', coord)
```

yardasol replied:
I updated your suggestion 

> 1. 100% valid (can surface sense be changes as a multiply operation?)
> 
> 2. too clever? Is it harder to understand for future developers?
> 
> 3. messing up surface naming conventions?

1. It's wonderful, but we can't multiply `int` by a `Surface` to get what we want.
2. I don't think so.
3. Nope.

I modified your original suggestion a bit so it works.

gonuke commented on openmc/model/funcs.py:
```diff
@@ -111,6 +112,151 @@ def borated_water(boron_ppm, temperature=293., pressure=0.1013, temp_unit='K',
+            fillet_upper = create_fillet(args, r, upper_fillet_radius, True)
+        if lower_fillet_radius > 0.:
+            fillet_lower = create_fillet(args, r, lower_fillet_radius, False)
+        if fillet_lower is not None and fillet_upper is not None:
+            fillet = fillet_lower | fillet_upper
+        elif fillet_lower is None:
+            fillet = fillet_upper
+        else:
+            fillet = fillet_lower
+
+        prism = prism & ~fillet
```
(Speculation?) If neither fillet is defined then `fillet` will be `None` in this last operation.  Is that geometry operation well defined if `fillet is None`?

If it is.... can you just use the `fillet_upper` and `fillet_lower` without the above logic to see if they are defined?

```suggestion
        prism = prism & ~(fillet_upper | fillet_lower)
```

Are all those operation well defined if one or both are `None`?

yardasol replied:
That's a great question. My guess is that it doesn't work that way based on the documentation for the `|` operator: 
```
class Union(Region, MutableSequence):
     r"""Union of two or more regions.
 
     Instances of Union are generally created via the | operator applied to two
     instances of :class:`openmc.Region`. This is illustrated in the following
     example:
 
     >>> s1 = openmc.ZPlane(z0=0.0)
     >>> s2 = openmc.Sphere(r=637.1e6)
     >>> type(-s2 | +s1)
     <class 'openmc.region.Union'>
 
     Instances of this class behave like a mutable sequence, e.g., they can be
     indexed and have an append() method.
 
     Parameters
     ----------
     nodes : iterable of openmc.Region

```

but maybe @paulromano or @pshriwise can verify?

yardasol replied:
> If neither fillet is defined then `fillet` will be `None` in this last operation. Is that geometry operation well defined if `fillet is None`?

At least one fillet is guaranteed to exist because this code block is contained within an if statement that requires at least one of `uppet_fillet_radius` and `lower_fillet_radius` to be nonzero.

gonuke reviewed (changes requested):
A few more ideas for streamlining the code.  I confess some are based on speculation of how things work and incomplete expertise on my part, but looked like opportunities for simplifying things.

yardasol commented on openmc/model/funcs.py:
```diff
@@ -111,6 +112,151 @@ def borated_water(boron_ppm, temperature=293., pressure=0.1013, temp_unit='K',
+    # Handle rounded corners if given
+    if upper_fillet_radius > 0. or lower_fillet_radius > 0.:
+        # Get torus class corresponding to given axis
+        tor = globals()['{}Torus'.format(axis.upper())]
```
Yes, and yes!

paulromano commented:
@yardasol Now I'm realizing that we already have [`openmc.model.RightCircularCylinder`](https://docs.openmc.org/en/latest/pythonapi/generated/openmc.model.RightCircularCylinder.html#openmc.model.RightCircularCylinder), which provides the same thing but without fillets. What do you think about just extending that class to support fillets rather than having two different ways to create cylindrical prisms?

yardasol commented:
> @yardasol Now I'm realizing that we already have [`openmc.model.RightCircularCylinder`](https://docs.openmc.org/en/latest/pythonapi/generated/openmc.model.RightCircularCylinder.html#openmc.model.RightCircularCylinder), which provides the same thing but without fillets. What do you think about just extending that class to support fillets rather than having two different ways to create cylindrical prisms?

This is a good idea. I've been busy doing other things, which is why I added the "on hold" tag, but hopefully I'll be able to get back to this soon (sometime this week most likely)

yardasol commented:
Finally got around to dealing with this one.

yardasol commented on docs/source/pythonapi/model.rst:
```diff
@@ -13,6 +13,7 @@ Convenience Functions
    openmc.model.borated_water
    openmc.model.hexagonal_prism
    openmc.model.rectangular_prism
+   openmc.model.cylindrical_prism
```
```suggestion
```

yardasol commented on openmc/model/surface_composite.py:
```diff
@@ -383,33 +387,183 @@ class RightCircularCylinder(CompositeSurface):
+
+                return cyl, torus, plane
+
+    def get_fillet(self):
+        upper_fillet = self.get_upper_fillet()
+        lower_fillet = self.get_lower_fillet()
+        has_upper_fillet = upper_fillet is not None
+        has_lower_fillet = lower_fillet is not None
+        if has_lower_fillet and has_upper_fillet:
+            fillet = lower_fillet | upper_fillet
+        elif has_upper_fillet and not has_lower_fillet:
+            fillet = upper_fillet
+        elif not has_upper_fillet and has_lower_fillet:
+            fillet = lower_fillet
+        else:
+            fillet = None
+        return fillet
+
+    def get_upper_fillet(self):
+        has_upper_fillet = hasattr(self, 'upper_fillet_plane')
+        if has_upper_fillet:
+            upper_fillet = +self.upper_fillet_cylinder & +self.upper_fillet_torus & +self.upper_fillet_plane
+        else:
+            upper_fillet = None
+        return upper_fillet
+
+    def get_lower_fillet(self):
+        has_lower_fillet = hasattr(self, 'lower_fillet_plane')
+        if has_lower_fillet:
+            lower_fillet = +self.lower_fillet_cylinder & +self.lower_fillet_torus & -self.lower_fillet_plane
+        else:
+            lower_fillet = None
+        return lower_fillet
```
```suggestion
    def _get_fillet(self):
        upper_fillet = self._get_upper_fillet()
        lower_fillet = self._get_lower_fillet()
        has_upper_fillet = upper_fillet is not None
        has_lower_fillet = lower_fillet is not None
        if has_lower_fillet and has_upper_fillet:
            fillet = lower_fillet | upper_fillet
        elif has_upper_fillet and not has_lower_fillet:
            fillet = upper_fillet
        elif not has_upper_fillet and has_lower_fillet:
            fillet = lower_fillet
        else:
            fillet = None
        return fillet

    def _get_upper_fillet(self):
        has_upper_fillet = hasattr(self, 'upper_fillet_plane')
        if has_upper_fillet:
            upper_fillet = +self.upper_fillet_cylinder & +self.upper_fillet_torus & +self.upper_fillet_plane
        else:
            upper_fillet = None
        return upper_fillet

    def _get_lower_fillet(self):
        has_lower_fillet = hasattr(self, 'lower_fillet_plane')
        if has_lower_fillet:
            lower_fillet = +self.lower_fillet_cylinder & +self.lower_fillet_torus & -self.lower_fillet_plane
        else:
            lower_fillet = None
        return lower_fillet
```

yardasol commented on openmc/model/surface_composite.py:
```diff
@@ -383,33 +387,183 @@ class RightCircularCylinder(CompositeSurface):
     def __neg__(self):
-        return -self.cyl & +self.bottom & -self.top
+        prism = -self.cyl & +self.bottom & -self.top
+        fillet = self.get_fillet()
```
```suggestion
        fillet = self._get_fillet()
```

yardasol commented on openmc/model/surface_composite.py:
```diff
@@ -383,33 +387,183 @@ class RightCircularCylinder(CompositeSurface):
     def __pos__(self):
-        return +self.cyl | -self.bottom | +self.top
+        prism = +self.cyl | -self.bottom | +self.top
+        fillet = self.get_fillet()
```
```suggestion
        fillet = _self.get_fillet()
```

yardasol commented:
@paulromano bump

paulromano commented on tests/unit_tests/test_geometry.py:
```diff
@@ -3,6 +3,7 @@
 
 import numpy as np
 import openmc
+import openmc.lib
```
```suggestion
```

paulromano reviewed (changes requested):
Sorry for the long delay @yardasol! Thanks for refactoring this into the RightCircularCylinder class -- it's looking really good

paulromano reviewed (approved):
Thanks for the quick updates @yardasol. Good to go!

paulromano commented:
@pshriwise do you have any further requests on this PR?

pshriwise commented:
> @pshriwise do you have any further requests on this PR?

No further requests from me 🚀

paulromano commented:
@pshriwise Thanks! need you to hit the approve button for this one to land

pshriwise reviewed (approved):
Thanks @yardasol!
